File: src/main/java/com/leijendary/spring/boot/template/api/v1/service/SampleTableService.java
package com.leijendary.spring.boot.template.api.v1.service;

import java.util.UUID;

import com.leijendary.spring.boot.template.api.v1.data.SampleRequest;
import com.leijendary.spring.boot.template.api.v1.data.SampleResponse;
import com.leijendary.spring.boot.template.api.v1.mapper.SampleMapper;
import com.leijendary.spring.boot.template.api.v1.search.SampleSearch;
import com.leijendary.spring.boot.template.core.data.QueryRequest;
import com.leijendary.spring.boot.template.core.exception.ResourceNotFoundException;
import com.leijendary.spring.boot.template.event.SampleProducer;
import com.leijendary.spring.boot.template.repository.SampleTableRepository;
import com.leijendary.spring.boot.template.specification.SampleListSpecification;

import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class SampleTableService {

    public static final String CACHE_KEY = "sample:v1";

    private static final String RESOURCE_NAME = "Sample Table";
    private static final SampleMapper MAPPER = SampleMapper.INSTANCE;

    private final SampleProducer sampleProducer;
    private final SampleSearch sampleSearch;
    private final SampleTableRepository sampleTableRepository;

    @Transactional(readOnly = true)
    public Page<SampleResponse> list(final QueryRequest queryRequest, final Pageable pageable) {
        final var specification = SampleListSpecification.builder()
                .query(queryRequest.getQuery())
                .build();

        return sampleTableRepository
                .findAll(specification, pageable)
                .map(MAPPER::toResponse);
    }

    @Transactional
    public SampleResponse create(final SampleRequest sampleRequest) {
        final var sampleTable = MAPPER.toEntity(sampleRequest);

        sampleTableRepository.save(sampleTable);

        // Save the object to elasticsearch
        sampleSearch.save(sampleTable);

        // Create an event object
        final var sampleEvent = MAPPER.toEvent(sampleTable);

        // Send the event object to kafka
        sampleProducer.create(sampleEvent);

        return MAPPER.toResponse(sampleTable);
    }

    @Transactional(readOnly = true)
    public SampleResponse get(final UUID id) {
        return sampleTableRepository.findById(id)
                .map(MAPPER::toResponse)
                .orElseThrow(() -> new ResourceNotFoundException(RESOURCE_NAME, id));
    }

    @Transactional
    public SampleResponse update(final UUID id, final SampleRequest sampleRequest) {
        final var sampleTable = sampleTableRepository.findLockedById(id)
                .orElseThrow(() -> new ResourceNotFoundException(RESOURCE_NAME, id));

        MAPPER.update(sampleRequest, sampleTable);

        sampleTableRepository.save(sampleTable);

        // Update the object from elasticsearch
        sampleSearch.update(sampleTable);

        // Create an event object
        final var sampleEvent = MAPPER.toEvent(sampleTable);

        // Send the event object to kafka
        sampleProducer.update(sampleEvent);

        return MAPPER.toResponse(sampleTable);
    }

    @Transactional
    public void delete(final UUID id) {
        final var sampleTable = sampleTableRepository.findLockedById(id)
                .orElseThrow(() -> new ResourceNotFoundException(RESOURCE_NAME, id));

        sampleTableRepository.softDelete(sampleTable);

        // Delete the object from elasticsearch
        sampleSearch.delete(id);

        // Create an event object
        final var sampleEvent = MAPPER.toEvent(sampleTable);

        // Send the event object to kafka
        sampleProducer.delete(sampleEvent);
    }

    @Transactional
    public void reindex() {
        sampleTableRepository.streamAll()
                .parallel()
                .forEach(sampleSearch::save);
    }
}

File: src/main/kotlin/com/leijendary/spring/boot/template/api/v1/service/SampleTableService.kt
package com.leijendary.spring.boot.template.api.v1.service

import com.leijendary.spring.boot.template.api.v1.data.SampleRequest
import com.leijendary.spring.boot.template.api.v1.data.SampleResponse
import com.leijendary.spring.boot.template.api.v1.mapper.SampleMapper
import com.leijendary.spring.boot.template.api.v1.search.SampleSearch
import com.leijendary.spring.boot.template.core.data.QueryRequest
import com.leijendary.spring.boot.template.core.data.Seek
import com.leijendary.spring.boot.template.core.data.Seekable
import com.leijendary.spring.boot.template.core.exception.ResourceNotFoundException
import com.leijendary.spring.boot.template.data.SampleCreateEvent
import com.leijendary.spring.boot.template.data.SampleDeleteEvent
import com.leijendary.spring.boot.template.data.SampleUpdateEvent
import com.leijendary.spring.boot.template.model.SampleTable
import com.leijendary.spring.boot.template.repository.SampleTableRepository
import com.leijendary.spring.boot.template.specification.SampleListSpecification
import org.springframework.context.ApplicationEventPublisher
import org.springframework.data.jpa.domain.Specification
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional
import java.util.*

@Service
class SampleTableService(
    private val applicationEventPublisher: ApplicationEventPublisher,
    private val sampleSearch: SampleSearch,
    private val sampleTableRepository: SampleTableRepository,
) {
    companion object {
        const val CACHE_NAME = "sample:v1"
        private val MAPPER: SampleMapper = SampleMapper.INSTANCE
        private val SOURCE = listOf("data", "SampleTable", "id")
    }

    @Transactional(readOnly = true)
    fun seek(queryRequest: QueryRequest, seekable: Seekable): Seek<SampleResponse> {
        val specification: Specification<SampleTable> = SampleListSpecification(queryRequest.query)

        return sampleTableRepository
            .findAll(SampleTable::class, specification, seekable)
            .map { MAPPER.toResponse(it) }
    }

    @Transactional
    fun create(sampleRequest: SampleRequest): SampleResponse {
        val sampleTable: SampleTable = MAPPER.toEntity(sampleRequest)

        sampleTableRepository.save(sampleTable)

        applicationEventPublisher.publishEvent(SampleCreateEvent(sampleTable))

        return MAPPER.toResponse(sampleTable)
    }

    @Transactional(readOnly = true)
    fun get(id: UUID): SampleResponse {
        return sampleTableRepository.findById(id)
            .map { MAPPER.toResponse(it) }
            .orElseThrow { ResourceNotFoundException(SOURCE, id) }
    }

    @Transactional
    fun update(id: UUID, sampleRequest: SampleRequest): SampleResponse {
        val sampleTable: SampleTable = sampleTableRepository.findLockedById(id)
            ?: throw ResourceNotFoundException(SOURCE, id)

        MAPPER.update(sampleRequest, sampleTable)

        sampleTableRepository.save(sampleTable)

        applicationEventPublisher.publishEvent(SampleUpdateEvent(sampleTable))

        return MAPPER.toResponse(sampleTable)
    }

    @Transactional
    fun delete(id: UUID) {
        val sampleTable: SampleTable = sampleTableRepository.findLockedById(id)
            ?: throw ResourceNotFoundException(SOURCE, id)

        sampleTableRepository.softDelete(sampleTable)

        applicationEventPublisher.publishEvent(SampleDeleteEvent(sampleTable))
    }

    @Transactional(readOnly = true)
    fun reindex() {
        sampleTableRepository.streamAll()
            .parallel()
            .forEach { sampleSearch.save(it) }
    }
}